### PR TITLE
Feature/enable layers

### PR DESCRIPTION
**Pull request by Bluesmile82**
Second Contextual-Layer PR 🏎 
(This PR contains the Feature/contextual-layers)
Also it has a big refactor extracting components and removing the map as its not indispensable
- [x] Add LayersSwitcher to toggle Layers visibility in the API
- [x] Add functionality for team Layers:
  - [x] Checkbox to add a team layer instead of a personal layer
  - [x] Display team layers with a label
  - [x] Add Custom layers
  - [x] Validation for custom layers

### CODE REVIEW

File: src/components/ui/Checkbox.js [hunk 1]
```diff
@@ -1,14 +1,19 @@
...
 import PropTypes from 'prop-types';
+import { injectIntl } from 'react-intl';
 
-function Checkbox({ id, callback, label, defaultChecked}) {
+function Checkbox({ id, callback, label, defaultChecked, checked , labelId, intl }) {
   return (
     <div className="c-checkbox">
       <div className="checkbox">
-        <input type="checkbox" id={id} onChange={callback} defaultChecked={defaultChecked}/>
+        { typeof checked !== 'undefined' ?
```
Comment (edbrett) [inline src/components/ui/Checkbox.js]: Can this not go directly into the checked={} param of the input?

File: src/components/map/Map.js [hunk 2]
```diff
@@ -52,7 +56,19 @@ class Map extends React.Component {
                       .addTo(this.map)
                       .setZIndex(0);
   }
-
+    
+  updateLayers(layerUrls) {
```
Comment (edbrett) [inline src/components/map/Map.js]: I still dont like this. this.map can easily be passed outside of the component to the Layers component and be added there. This is poluting the map component and adding complexity when it is there to be reusable. All of this code should be inside the component that manages the layers.

File: src/components/pages/settings/LayersForm.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  render() {
+    return (
+      <div className="c-layers-form">
+        <div className="title"><FormattedMessage id={"settings.contextualLayers"} /></div>
+        <Tab 
+          options={["settings.gfwLayers", "settings.customLayers"]}
+          selectedIndex={this.state.tabIndex}
+          handleTabIndexChange={this.handleTabIndexChange}
+        />
+        <form onSubmit={(e) => this.addLayers(e)}>
```
Comment (edbrett) [inline src/components/pages/settings/LayersForm.js]: Why not use the Form component we have been using for other forms?

File: src/components/pages/settings/LayersForm.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+          <Card className={"-big"}>
+            {this.state.tabIndex === 0 ? 
+              this.state.GFWLayers && this.state.GFWLayers.map((GFWlayer, i) => 
+                <div key={i}>
```
Comment (edbrett) [inline src/components/pages/settings/LayersForm.js]: The key could go on the Checkbox components then you wouldn't need these divs

File: src/components/pages/settings/LayersForm.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                  id={`${i}${GFWlayer.title}`} 
+                  label={ GFWlayer.title }
+                  callback={() => this.toggleGFWLayer(GFWlayer)} 
+                  checked={ GFWlayer.enabled || false }
+                />
+                </div>
+              ) : null
+            }
+          </Card> 
+          <div>
```
Comment (edbrett) [inline src/components/pages/settings/LayersForm.js]: these divs aren't needed

File: src/components/pages/settings/LayersManager.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+
+import LayersForm from './LayersFormContainer';
+import LayersSwitcher from './LayersSwitcherContainer';
+
+class LayersManager extends React.Component {  
+  render() {
+    return (
+      <div className="c-layers-manager">
```
Comment (edbrett) [inline src/components/pages/settings/LayersManager.js]: This is interesting, and also not particularly important, but I generally use .l-class as opposed to .c-class when something contains lots of other components. Layout classes handle positioning on the page etc, and .c-classes handle the finer details. Good to follow the same format through our projects so would be good to decide on this.

File: src/components/pages/settings/LayersFormContainer.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import { connect } from 'react-redux';
+import LayersForm from './LayersForm';
+import { getGFWLayers, createLayer } from '../../../modules/layers';
+
+
+const mapStateToProps = ({ layers, teams }) => {
+    return {
```
Comment (edbrett) [inline src/components/pages/settings/LayersFormContainer.js]: This again is point of preference, but as there are not a lot of props, and similar things are getting passed to LayerForm and LayerSwitcher, we could have a single Container to the LayersForm (their parent), and then pass the need props to each. This leaves the Switcher and Form as UI components and you can control the state of each more clearly. Lets discuss.

File: src/components/pages/settings/LayersSwitcher.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
+
+import { FormattedMessage } from 'react-intl';
+import SwitchButton from 'react-switch-button';
+
+class LayersSwitcher extends React.Component {
+  componentWillMount(){
+    this.props.getLayers();
```
Comment (edbrett) [inline src/components/pages/settings/LayersSwitcher.js]: Should this be called on the parent (LayersManager)?

File: src/components/pages/settings/LayersSwitcher.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  toggleLayer = (layer) => {
+    this.props.toggleLayer(layer, !layer.attributes.enabled);
+  }
+
+  render() {
+    return (
+      <div>
+        <div className="title"><FormattedMessage id={"settings.selectedLayers"} /></div>
```
Comment (edbrett) [inline src/components/pages/settings/LayersSwitcher.js]: could use a h3 tag here to be more expressive.

File: src/components/pages/settings/LayersSwitcher.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        <div className="section">
+          {this.props.selectedLayers.map((selectedLayer, i) => (
+            <div key={i}>
```
Comment (edbrett) [inline src/components/pages/settings/LayersSwitcher.js]: Again can place these on the component to reduce code.

File: src/components/pages/settings/Settings.js [hunk 11]
```diff
@@ -31,67 +32,19 @@ class Settings extends React.Component {
         : 
           <Hero title={"settings.name"} />
         }
-        {loading ? 
-          null
-          :
-          <div>
-          {(team && !editing) ?
+        <div className="l-content">
```
Comment (edbrett) [inline src/components/pages/settings/Settings.js:37]: Nice work. Big reduction and simplification.

File: src/components/pages/settings/TeamsForm.js [hunk 12]
```diff
@@ -145,8 +145,24 @@ class TeamsForm extends React.Component {
             <div className="row small-12 columns">
               <div className="c-form -nav">
                 { team &&
-                  <div><button type="button" onClick={this.handleCancel} disabled={this.props.loading} className={`c-button -light `}><FormattedMessage id={"common.cancel"} /></button></div>}
-                <div><button type="submit" disabled={this.props.loading} className="c-button -right"><FormattedMessage id={"common.save"} /></button></div>
+                  <div>
```
Comment (edbrett) [inline src/components/pages/settings/TeamsForm.js]: Empty divs again. Not needed.

File: src/components/pages/settings/TeamsForm.js [hunk 13]
```diff
@@ -145,8 +145,24 @@ class TeamsForm extends React.Component {
...
+                    <button 
+                      type="button" 
+                      onClick={this.handleCancel} 
+                      disabled={this.props.loading} 
+                      className={`c-button -light `}>
+                        <FormattedMessage id={"common.cancel"} />
+                    </button>
+                  </div>
+                  }
+                <div>
```
Comment (edbrett) [inline src/components/pages/settings/TeamsForm.js]: Here too.

File: src/components/ui/Tab.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { FormattedMessage } from 'react-intl';
+
+function Tab({ options, selectedIndex, handleTabIndexChange }) {
```
Comment (edbrett) [inline src/components/ui/Tab.js:5]: Cooooool. 😎 . Lets not forget we have this for other projects.

File: src/styles/_react-switch-button.scss [hunk 15]
```diff
@@ -0,0 +1,229 @@
+/**
+ * React Switch Button control style.
```
Comment (edbrett) [inline src/styles/_react-switch-button.scss:2]: Not a big fan of copying styles to override as often we only need to change a few classes. But if it works 👍

File: src/styles/components/_c-layers-manager.scss [hunk 16]
```diff
@@ -0,0 +1,11 @@
+.c-layers-manager {
+    padding-top: 60px;
+  .section {
```
Comment (edbrett) [inline src/styles/components/_c-layers-manager.scss]: spacing.

File: src/styles/components/_c-map.scss [hunk 17]
```diff
@@ -1,3 +1,9 @@
 .c-map {
   height: calc(100vh - #{$nav-height} - #{$hero-height} - #{$footer-height});
-}
+  &.-layers-container {
```
Comment (edbrett) [inline src/styles/components/_c-map.scss:3]: spacing

**Review by edbrett — CHANGES_REQUESTED** [review]
I am still getting problems with using the layer manager in general on my machine. Lets review when you get to office together.

**Comment by edbrett** [comment]
Also can we organise the settings pages better. There are way to many files in there and it is hard to know what are parents and what are child components. We can discuss when you get to office. We have been following a structure for this for other pages.

**Comment by edbrett** [comment]
Also also, I think as some of these components are starting to become larger (and manage other components) we should be diving up the functions with comments to make is clearer what is happening.

File: src/components/pages/settings/LayersSwitcher.js (hunk 10 above)
Comment (edbrett) [reply src/components/pages/settings/LayersSwitcher.js]: Sorry this is wrong. Ignore my comment

File: src/components/ui/Checkbox.js (hunk 1 above)
Comment (Bluesmile82) [reply src/components/ui/Checkbox.js]: We don't want checked to be false if the checked is not defined. We don't want it to have the param. So the checkbox is still uncontrolled.

File: src/styles/_base.scss [hunk 18]
```diff
@@ -52,3 +52,11 @@ textarea:focus,
 button:focus {
   outline: none;
 }
+
+h3 {
```
Comment (edbrett) [inline src/styles/_base.scss]: we have a texts.scss file for things like this as base should be to normalize the css.

File: src/styles/_base.scss [hunk 19]
```diff
@@ -52,3 +52,11 @@ textarea:focus,
...
+  font-weight: bold;
```
Comment (edbrett) [inline src/styles/_base.scss]: we have a settings file for things like bold.

Comment (edbrett) [reply src/styles/_base.scss]: And always use number for font weights as font families etc may not do what you expect with bold.

**Pull request merged**